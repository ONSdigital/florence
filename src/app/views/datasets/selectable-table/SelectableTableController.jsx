import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Link } from 'react-router';
import dateFormat from 'dateformat';

import url from '../../../utilities/url';

const propTypes = {
    values: PropTypes.arrayOf(PropTypes.shape({
        title: PropTypes.string.isRequired,
        date: PropTypes.string,
        id: PropTypes.string.isRequired,
        instances: PropTypes.arrayOf(PropTypes.shape({
            date: PropTypes.string,
            edition: PropTypes.string.isRequired,
            version: PropTypes.oneOfType([
                PropTypes.string.isRequired,
                PropTypes.number.isRequired
            ])
        }))
    }))
}

export default class SelectableTableController extends Component {
    constructor(props) {
        super(props);
    }

    renderDatasetInstances(instances) {
        return instances.map((instance, i) => {
            if (instance !== undefined) {
                return (
                    <div key={i} className="grid simple-table__row">
                        <div className="grid__col-4">
                            {dateFormat(instance.date, "HH:MM:ss dd/mm/yy")}
                        </div>
                        <div className="grid__col-2">
                            {instance.edition}
                        </div>
                        <div className="grid__col-4">
                            {instance.version}
                        </div>
                        <div className="grid__col-2">
                            {/* Wrap in span to stop grid col from forcing link to be full width 
                            (which made focus outline much wider than the link itself) */}
                            <span><Link to={instance.url}>View</Link></span>
                        </div>
                    </div>
                )
            }
        })
    }

    render() {
        return (
            <div className="selectable-table">
                    <div className="selectable-table__heading">
                        Title
                    </div>
                    {this.props.values.map((item, index) => {
                        return (
                            <details key={index} className="selectable-table__details">
                                <summary className="selectable-table__summary">
                                    <div className="grid">
                                        <div className="grid__col-6">
                                            <strong>{item.title}</strong>
                                        </div>
                                        <div className="grid__col-6">
                                            {item.date}
                                        </div>
                                    </div>
                                </summary>
                                <div>
                                    <Link className="inline-block margin-bottom--1" to={url.resolve(`datasets/${item.id}/metadata`)}> Edit dataset details</Link>
                                    {item.instances.length > 0 &&
                                        <div>
                                            <p className="font-weight--400 grid__col-12">
                                                New versions:
                                            </p>
                                            <div className="grid simple-table__heading">
                                                <div className="grid__col-4">
                                                    Date
                                                </div>
                                                <div className="grid__col-2">
                                                    Edition
                                                </div>
                                                <div className="grid__col-2">
                                                    Version
                                                </div>
                                            </div>
                                            { this.renderDatasetInstances(item.instances) }
                                        </div>
                                    }
                                </div>
                            </details>
                        )
                    })}
            </div>
        )
    }
}

SelectableTableController.propTypes = propTypes;
