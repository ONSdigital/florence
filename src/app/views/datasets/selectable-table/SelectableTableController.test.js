import React from 'react';
import SelectableTableController from './SelectableTableController.jsx';
import renderer from 'react-test-renderer';

jest.mock('../../../utilities/url', () => {
    return {
        resolve: function() {
            //
        },
        parent: function() {}
    }
});

const values = [
    {
        title: "CPI",
        date: "17-10-2017",
        datasetURL: "/florence/datasets/12345/metadata",
        instances: [
            {
                date: "Tue Oct 17 2017 17:11:59",
                edition: "2016",
                version: "-",
                url: "/florence/datasets/12345/instances/6789/metdata"
            }
        ],
        id: "meeeeeh",
        status: "completed"
    }
]

test('Selectable table with values renders component with all values', () => {
    const component = renderer.create(
        <SelectableTableController 
            values={values}
        />
    );

    expect(component.toJSON()).toMatchSnapshot();
});