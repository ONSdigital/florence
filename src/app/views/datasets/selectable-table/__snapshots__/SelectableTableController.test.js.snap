// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Selectable table with values renders component with all values 1`] = `
<div
  className="selectable-table"
>
  <div
    className="selectable-table__heading"
  >
    Title
  </div>
  <details
    className="selectable-table__details"
  >
    <summary
      className="selectable-table__summary"
    >
      <div
        className="grid"
      >
        <div
          className="grid__col-6"
        >
          <strong>
            CPI
          </strong>
        </div>
        <div
          className="grid__col-6"
        >
          17-10-2017
        </div>
      </div>
    </summary>
    <div>
      <a
        className="inline-block margin-bottom--1"
        onClick={[Function]}
        style={Object {}}
      >
         Edit dataset details
      </a>
      <div>
        <p
          className="font-weight--400 grid__col-12"
        >
          New versions:
        </p>
        <div
          className="grid simple-table__heading"
        >
          <div
            className="grid__col-4"
          >
            Date
          </div>
          <div
            className="grid__col-2"
          >
            Edition
          </div>
          <div
            className="grid__col-2"
          >
            Version
          </div>
        </div>
        <div
          className="grid simple-table__row"
        >
          <div
            className="grid__col-4"
          >
            17:11:59 17/10/17
          </div>
          <div
            className="grid__col-2"
          >
            2016
          </div>
          <div
            className="grid__col-4"
          >
            -
          </div>
          <div
            className="grid__col-2"
          >
            <span>
              <a
                onClick={[Function]}
                style={Object {}}
              >
                View
              </a>
            </span>
          </div>
        </div>
      </div>
    </div>
  </details>
</div>
`;
